File: assignment 5/python/programs/6.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None


class DoublyCircularLinkedList:
    def __init__(self):
        self.head = None

    # Insert a node at the beginning of the doubly circular linked list
    def insert_at_beginning(self, data):
        new_node = Node(data)
        if self.head is None:
            new_node.next = new_node
            new_node.prev = new_node
            self.head = new_node
        else:
            last_node = self.head.prev
            new_node.next = self.head
            new_node.prev = last_node
            self.head.prev = new_node
            last_node.next = new_node
            self.head = new_node

    # Insert a node at the end of the doubly circular linked list
    def insert_at_end(self, data):
        new_node = Node(data)
        if self.head is None:
            new_node.next = new_node
            new_node.prev = new_node
            self.head = new_node
        else:
            last_node = self.head.prev
            new_node.next = self.head
            new_node.prev = last_node
            self.head.prev = new_node
            last_node.next = new_node

    # Insert a node after a specific element
    def insert_after(self, data, key):
        new_node = Node(data)
        if self.head is None:
            print("List is empty. Cannot perform insertion.")
            return
        current = self.head
        while current:
            if current.data == key:
                new_node.next = current.next
                current.next.prev = new_node
                current.next = new_node
                new_node.prev = current
                return
            current = current.next
            if current == self.head:
                print(f"Element {key} not found in the list.")
                return
        print(f"Element {key} not found in the list.")

    # Delete a node by value
    def delete(self, data):
        if self.head is None:
            print("List is empty. Cannot perform deletion.")
            return
        current = self.head
        while True:
            if current.data == data:
                if current.next == self.head:
                    if current.prev == current:
                        self.head = None
                    else:
                        current.prev.next = self.head
                        self.head.prev = current.prev
                        self.head = current.next
                else:
                    current.prev.next = current.next
                    current.next.prev = current.prev
                return
            current = current.next
            if current == self.head:
                print(f"Element {data} not found in the list.")
                return

    # Display the doubly circular linked list
    def display(self):
        if self.head is None:
            print("Doubly Circular Linked List is empty.")
            return

        current = self.head
        while True:
            print(current.data, end=" <-> ")
            current = current.next
            if current == self.head:
                break
        print("Head")


if __name__ == "__main__":
    doubly_circular_linked_list = DoublyCircularLinkedList()

    # Insert elements into the doubly circular linked list
    doubly_circular_linked_list.insert_at_end(1)
    doubly_circular_linked_list.insert_at_end(2)
    doubly_circular_linked_list.insert_at_end(3)

    # Display the doubly circular linked list
    print("Doubly Circular Linked List:")
    doubly_circular_linked_list.display()

    # Insert an element after a specific element
    doubly_circular_linked_list.insert_after(4, 2)

    # Display the doubly circular linked list after insertion
    print("Doubly Circular Linked List after Insertion:")
    doubly_circular_linked_list.display()

    # Delete an element from the doubly circular linked list
    doubly_circular_linked_list.delete(2)

    # Display the doubly circular linked list after deletion
    print("Doubly Circular Linked List after Deletion:")
    doubly_circular_linked_list.display()

File: assignment 5/c/programs/6.c
#include <stdio.h>
#include <stdlib.h>

// Define the structure for a node in the doubly circular linked list
struct Node
{
    int data;
    struct Node *prev;
    struct Node *next;
};

// Function to create a new node
struct Node *createNode(int data)
{
    struct Node *newNode = (struct Node *)malloc(sizeof(struct Node));
    if (newNode == NULL)
    {
        printf("Memory allocation failed.\n");
        exit(1);
    }
    newNode->data = data;
    newNode->prev = NULL;
    newNode->next = NULL;
    return newNode;
}

// Function to insert a node at the end of the doubly circular linked list
struct Node *insertAtEnd(struct Node *head, int data)
{
    struct Node *newNode = createNode(data);

    if (head == NULL)
    {
        newNode->next = newNode;
        newNode->prev = newNode;
        return newNode;
    }

    newNode->next = head;
    newNode->prev = head->prev;
    head->prev->next = newNode;
    head->prev = newNode;

    return head;
}

// Function to delete a node at a specific position in the doubly circular linked list
struct Node *deleteAtPosition(struct Node *head, int position)
{
    if (head == NULL)
    {
        printf("List is empty.\n");
        return NULL;
    }

    if (position <= 0)
    {
        printf("Invalid position.\n");
        return head;
    }

    if (position == 1)
    {
        struct Node *temp = head;

        if (head->next == head)
        {
            free(head);
            return NULL; // List becomes empty
        }

        head->prev->next = head->next;
        head->next->prev = head->prev;
        head = head->next;

        free(temp);
        return head;
    }

    struct Node *current = head;
    int count = 1;

    // Find the node to be deleted
    while (count < position && current->next != head)
    {
        current = current->next;
        count++;
    }

    if (count < position)
    {
        printf("Position out of range.\n");
        return head;
    }

    current->prev->next = current->next;
    current->next->prev = current->prev;

    free(current);
    return head;
}

// Function to display the doubly circular linked list
void display(struct Node *head)
{
    if (head == NULL)
    {
        printf("List is empty.\n");
        return;
    }

    struct Node *current = head;

    do
    {
        printf("%d -> ", current->data);
        current = current->next;
    } while (current != head);

    printf("\n");
}

int main()
{
    struct Node *head = NULL;

    // Initialize a predefined doubly circular linked list
    head = insertAtEnd(head, 10);
    head = insertAtEnd(head, 20);
    head = insertAtEnd(head, 30);

    printf("Initial Doubly Circular Linked List: ");
    display(head);

    int data, position;

    // Insertion
    printf("\nEnter data to insert at the end: ");
    scanf("%d", &data);
    head = insertAtEnd(head, data);

    // Display the list after insertion
    printf("Updated Doubly Circular Linked List: ");
    display(head);

    // Deletion
    printf("\nEnter position to delete: ");
    scanf("%d", &position);
    head = deleteAtPosition(head, position);

    // Display the list after deletion
    printf("Final Doubly Circular Linked List: ");
    display(head);

    return 0;
}
